dts/msm8974/msm8974-samsung-r14.dts

```diff
@@ -13,4 +13,23 @@
 		compatible = "samsung,klte", "qcom,msm8974", "lk2nd,device";
 		lk2nd,match-bootloader = "G900F*";
 	};
+
+	/*
+	 samsung-klte has at least 2 known panel variations:
+	   * S6E3FA2 (mdss_mdp.panel=1:dsi:0:mdss_dsi_samsung_1080p_cmd_fa2)
+	   * EA8064G (mdss_mdp.panel=1:dsi:0:mdss_dsi_samsung_1080p_cmd_mag)
+	 Upstream mainline kernel already has "samsung,s6e3fa2" compatible set
+	   for panel in dts, and we'll not change that. We'll only do
+	   replacement for ea8064g one (replace samsung,s6e3fa2 to
+	   samsung,ea8064g).
+	*/
+	panel {
+		/* Default is Samsung S6E3FA2 */
+		compatible = "samsung,s6e3fa2";
+
+		/* MagnaChip EA8064G */
+		mdss_dsi_samsung_1080p_cmd_mag {
+			compatible = "samsung,ea8064g";
```

Comment:
Note that due to the way the panel detection is implemented at the moment this will effectively result in `compatible = "samsung,ea8064g", "samsung,s6e3fa2";`. When the kernel is missing support for `ea8064g`, it would still use the `s6e3fa2` driver even if it is known that it is the wrong one.

This works well for the devices with "generic" compatible in lk2nd, because there it results in something like `compatible = "samsung,s6d78a0-gh9607501a", "samsung,gprime-panel"`. Personally I prefer the approach with the generic compatible because it ensures that users never accidentally run the wrong panel driver, which could, theoretically, cause permanent hardware damage (although I have no heard of such a case so far).

But I'll leave this up to you I think :)

Reply:
First, the kernel will have support for both variations anyway.

Second, panel driver checks panel's MTP ID and refuses to power on unknown panel. Permanent hw damage is not possible.

Third, what? So it is not replacing, it is only prepending? This doesn't seem right, we need a mechanism to fully replace the compatible string. I do not think "generic compatible" is upstreamable right now. Maybe I can add handling for new property like replace-compatible = "xxx" ...?

Though, `compatible = "samsung,ea8064g", "samsung,s6e3fa2";` should still work for my use case. But I couldn't confirm that compatible string is changed at all during my brief test yesterday.

Reply:
I'm not sure I've got all bloody C pointer arithmetics right, I meant something like this: https://github.com/msm8916-mainline/lk2nd/pull/205/commits/e67e5232757743a2584e3b3c1beeabc58881a608

Reply:
> First, the kernel will have support for both variations anyway.

Not necessarily I think, what if someone boots older MSM8974 releases with these lk2nd changes? It will set the `"samsung,ea8064g"` compatible but the old kernel does not have a driver for it.

> Second, panel driver checks panel's MTP ID and refuses to power on unknown panel. Permanent hw damage is not possible.

That is a good point. Note that most other panel drivers do not have such checks though.

> Third, what? So it is not replacing, it is only prepending? This doesn't seem right, we need a mechanism to fully replace the compatible string.

I'm not opposed to add support to replace the compatible string completely, with something like your proposed patch. I'm not sure why the existing code does not seem right to you though: it works exactly as intended. It just does not fit exactly to your specific requirements. :)

> I do not think "generic compatible" is upstreamable right now. 

Why? My thought behind it is the following:

It is a well-established convention that device-specific details are left up to the bootloader to fill in. Take `/memory` for instance: It is usually set to dummy values (0) by default and the bootloader is expected to fill in the amount of RAM at runtime. We could hardcode the most common amount of RAM but we usually don't, perhaps because this increases the chance that users run an incorrect device tree and get confused.

The generic panel compatible is not just a dummy value though. It is kept by lk2nd (by prepending the more specific compatible) because I argue that it has semantic value. Consider the following example:

The device tree source has:

```dts
compatible = "samsung,gprime-panel";
```

This describes exactly what we know for sure at compile time: The device has one of the possible panels of the `samsung,gprime` device. We don't know which panel exactly.

Having this generic compatible might seem useless because no driver in Linux will bind to it. But this is just one possible implementation: Theoretically, you could write a generic driver for `"samsung,gprime-panel"` by implementing the panel detection directly in Linux or another operating system.

When using lk2nd, this compatible is changed at runtime to:

```dts
compatible = "samsung,s6d78a0-gh9607501a", "samsung,gprime-panel";
```

Semantically this means: The panel is compatible with `samsung,s6d78a0-gh9607501a`, and more in general also `samsung,gprime-panel`.

Now the operating system (e.g. Linux) is relieved of doing the panel detection itself. Linux just binds a driver directly to `samsung,s6d78a0-gh9607501a`. If the specific panel is not yet supported by any driver in Linux then the panel is automatically disabled at runtime.

However, an operating system with the theoretic generic driver from above could still work correctly. It would ignore the `"samsung,s6d78a0-gh9607501a"` compatible and just binds the driver to `"samsung,gprime-panel"`.

This is quite similar to this example from the device tree specification:

> `compatible = "fsl,mpc8641", "ns16550";`
> In this example, an operating system would first try to locate a device driver that supported fsl,mpc8641. If a driver was not found, it would then try to locate a driver that supported the more general ns16550 device type.

Reply:
>  what if someone boots older MSM8974 releases with these lk2nd changes? It will set the "samsung,ea8064g" compatible but the old kernel does not have a driver for it.

It does have a driver for it. S6E3FA2 panel driver in older releases supports both `samsung,s6e3fa2` [1] and `samsung,ea8064g` compatibles. It ignores the difference and reads an actual MTP ID using single-byte read commands [2]. Then it checks is MTP ID is among supported ones, and only then operates accordingly.

In new releases drivers will support each their own compatible, but will still check MTP ID and refuse to work on unknown hardware.

I thnk, safety should be pretty okay here, hardware damage should be prevented in all cases.

[1] https://github.com/msm8974-mainline/linux/blob/qcom-msm8974-5.19.y/drivers/gpu/drm/panel/panel-samsung-s6e3fa2-full.c#L1158
[2] https://github.com/msm8974-mainline/linux/blob/qcom-msm8974-5.19.y/drivers/gpu/drm/panel/panel-samsung-s6e3fa2-full.c#L763

Reply:
> Note that most other panel drivers do not have such checks though.

I'm not suggesting you have to switch all devices to new mechanism. Can klte at least or now have a specific compatible, especially given the fact that it is already upstream in dts? _(And it's corresponding panel driver is not upstream, because I need to rewrite it before I send it again, and for this I need panel selection for klte to work in lk2nd)_



>>   I do not think "generic compatible" is upstreamable right now.

> Why? My thought behind it is the following:

What you're saying makes some sense to me, but I'll actually believe it's upstreamable the moment I see something like `compatible = "samsung,a5lte-panel"` in torvalds/linux repo. Until then it's only a theory :)

> It is a well-established convention that device-specific details are left up to the bootloader to fill in.

Right, but in this case we know exactly which specific panel is inside the device, so we can fill compatible property with exact specific value, so why do we even need generic compatible then? :)

Also, you've probably seen this recently: https://www.spinics.net/lists/linux-input/msg82097.html
```
>    compatible:
>      enum:
> +      - imagis,ist3038
>        - imagis,ist3038c
> +      - imagis,ist30xxb

No wildcards. Real model numbers.
```
Clearly, specific compatibles are preferred over generic or "wildcard" ones.

---

With all that said, I think, prepending specific compatible should also work for my use case, just because that's how compatibles work, right? So in case `compatible = "samsung,ea8064g", "samsung,s6e3fa2";` ea8064g driver will be tried first because it comes first. It's not safer to "fallback" to fa2 driver, it's safe and backwards compatible in both cases. In this case of prepending just a few extra of bytes of memory will be wasted in fdt blob :)